core/src/repair_service.rs

```diff
@@ -238,19 +286,212 @@ impl RepairService {
+                // If the slot is full, no further need to repair this slot
+                None
+            } else {
+                Some(Self::generate_repairs_for_slot(
+                    blockstore,
+                    slot,
+                    &slot_meta,
+                    MAX_REPAIR_PER_DUPLICATE,
+                ))
+            }
+        } else {
+            error!("Slot meta for duplicate slot does not exist, cannot generate repairs");
+            // Filter out this slot from the set of duplicates to be repaired as
+            // the SlotMeta has to exist for duplicates to be generated
+            None
+        }
+    }
+
+    fn generate_and_send_duplicate_repairs(
+        duplicate_slot_repair_statuses: &mut HashMap<Slot, DuplicateSlotRepairStatus>,
+        cluster_slots: &ClusterSlots,
+        blockstore: &Blockstore,
+        serve_repair: &ServeRepair,
+        repair_stats: &mut RepairStats,
+        repair_socket: &UdpSocket,
+    ) {
+        duplicate_slot_repair_statuses.retain(|slot, status| {
+            Self::update_duplicate_slot_repair_addr(*slot, status, cluster_slots, serve_repair);
+            if let Some(repair_addr) = status.repair_addr {
+                let repairs = Self::generate_duplicate_repairs_for_slot(&blockstore, *slot);
+
+                if let Some(repairs) = repairs {
+                    for repair_type in repairs {
+                        if let Err(e) = Self::serialize_and_send_request(
+                            &repair_type,
+                            repair_socket,
+                            &repair_addr,
+                            serve_repair,
+                            repair_stats,
+                        ) {
+                            info!("repair req send_to({}) error {:?}", repair_addr, e);
+                        }
+                    }
+                    true
+                } else {
+                    false
+                }
+            } else {
+                false
+            }
+        })
+    }
+
+    fn serialize_and_send_request(
+        repair_type: &RepairType,
+        repair_socket: &UdpSocket,
+        to: &SocketAddr,
+        serve_repair: &ServeRepair,
+        repair_stats: &mut RepairStats,
+    ) -> Result<()> {
+        let req = serve_repair.map_repair_request(&repair_type, repair_stats)?;
+        repair_socket.send_to(&req, to)?;
+        Ok(())
+    }
+
+    fn update_duplicate_slot_repair_addr(
+        slot: Slot,
+        status: &mut DuplicateSlotRepairStatus,
+        cluster_slots: &ClusterSlots,
+        serve_repair: &ServeRepair,
+    ) {
+        let now = timestamp();
+        if status.repair_addr.is_none()
+            || now.saturating_sub(status.start) >= MAX_DUPLICATE_WAIT_MS as u64
+        {
+            let repair_addr =
+                serve_repair.repair_request_duplicate_compute_best_peer(slot, cluster_slots);
+            status.repair_addr = repair_addr.ok();
```

Comment:
I think this should be `.unwrap()` or `.expect()` when `now.saturating_sub(status.start) >= MAX_DUPLICATE_WAIT_MS ` and `status.repair_addr.is_some()`

Otherwise, I think [these purging logic](https://github.com/solana-labs/solana/pull/9805/files#diff-f03a09b141837083bc32c6373631d89fR413) can be called multiple times. That's because if we set `status.reparit_addr()` to `None` here, following call of [this](https://github.com/solana-labs/solana/pull/9805/files#diff-f03a09b141837083bc32c6373631d89fR344) will return `false` to the `.retain()`. Then at the next loop, `find_new_duplicate_slots` will recreate and return the same `DuplicateSlotRepairStatus`, and finally `process_new_duplicate_slots` do the purging logic again.

First of all, rather a basic question, but, subsequent calls of `repair_request_duplicate_compute_best_peer(slot, cluster_slots)` can be `Err` once after we get `Ok`?

Or, [all of these purging operations](https://github.com/solana-labs/solana/pull/9805/files#diff-f03a09b141837083bc32c6373631d89fR413-R436) are idempotent and safe to retry? I haven't checked all. But it seems it mostly so like [this](https://github.com/solana-labs/solana/pull/9805/files#diff-92c739d9ad61135b886d1a44957fe485R87).

Anyway, this might be a bit fragile. And maintaining idempotency will be hard thing to do without testing.  Also, should we comment that we must ensure idempotency around [here](https://github.com/solana-labs/solana/pull/9805/files#diff-f03a09b141837083bc32c6373631d89fR413) if repeated purging is expected behavior?

Reply:
> I think this should be `.unwrap()` or `.expect()` when `now.saturating_sub(status.start) >= MAX_DUPLICATE_WAIT_MS ` and `status.repair_addr.is_some()`
> Otherwise, I think [these purging logic](https://github.com/solana-labs/solana/pull/9805/files#diff-f03a09b141837083bc32c6373631d89fR413) can be called multiple times. That's because if we set `status.reparit_addr()` to `None` here, following call of [this](https://github.com/solana-labs/solana/pull/9805/files#diff-f03a09b141837083bc32c6373631d89fR344) will return `false` to the `.retain()`. Then at the next loop, `find_new_duplicate_slots` will recreate and return the same `DuplicateSlotRepairStatus`, and finally `process_new_duplicate_slots` do the purging logic again.
> Good catch! so if the `repair_addr.is_none()`, `retain` should return true, fixed here: https://github.com/solana-labs/solana/pull/9805/commits/ca6d066242ec6c33df75847bb600f0974a6ad169#diff-f03a09b141837083bc32c6373631d89fR364, and added to the the test: https://github.com/solana-labs/solana/pull/9805/commits/ca6d066242ec6c33df75847bb600f0974a6ad169#diff-f03a09b141837083bc32c6373631d89fR364-R931

> First of all, rather a basic question, but, subsequent calls of `repair_request_duplicate_compute_best_peer(slot, cluster_slots)` can be `Err` once after we get `Ok`?
> Yeah! so gossip peers can disappear, so you could potentially get an error if there are no more gossip peers(unlikely), or more likely, there are no more peers that have the completed version of the slot, in which case it will return `None`.

> Or, [all of these purging operations](https://github.com/solana-labs/solana/pull/9805/files#diff-f03a09b141837083bc32c6373631d89fR413-R436) are idempotent and safe to retry? I haven't checked all. But it seems it mostly so like [this](https://github.com/solana-labs/solana/pull/9805/files#diff-92c739d9ad61135b886d1a44957fe485R87).
> Yeah, the idea is it should be safe to retry!
> Anyway, this might be a bit fragile. And maintaining idempotency will be hard thing to do without testing. Also, should we comment that we must ensure idempotency around [here](https://github.com/solana-labs/solana/pull/9805/files#diff-f03a09b141837083bc32c6373631d89fR413) if repeated purging is expected behavior?
> Yeah, the design was to support repeated purging. The local cluster test with the `FailEntryVerification` broadcast type actually tests repeated purging. This is because until this: https://github.com/solana-labs/solana/pull/9805/files#diff-fbd4fb46535186d22af03b84dcd08958R51-R55 runs, it will continually serve bad repair requests which will cause the slot to be marked dead over and over. Then this slot will be detected as dead (see https://github.com/solana-labs/solana/pull/9805/files#diff-f03a09b141837083bc32c6373631d89fR452-R460) which will lead to repeated purging until a good repair can eventually be made

Reply:
Oh, i see! thanks for explaining!